File: sdk/typescript/src/pkce.ts
import type { PKCEPair } from './types.js';

/**
 * Convierte un ArrayBuffer o Uint8Array a una cadena Base64URL sin padding.
 */
function toBase64Url(bytes: Uint8Array): string {
  let binary = '';
  const len = bytes.byteLength;
  for (let i = 0; i < len; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  const base64 = globalThis.btoa ? globalThis.btoa(binary) : Buffer.from(bytes).toString('base64');
  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/**
 * Genera un par criptográfico PKCE (Code Verifier y Code Challenge S256).
 * Compatible con Node.js 18+, Bun, Deno, navegadores y Edge runtime (Web Crypto API).
 */
export async function generatePKCE(length: number = 64): Promise<PKCEPair> {
  if (length < 43 || length > 128) {
    throw new Error('La longitud del code_verifier debe estar entre 43 y 128 caracteres');
  }

  const cryptoObj = globalThis.crypto;
  if (!cryptoObj || !cryptoObj.subtle) {
    throw new Error('Web Crypto API no disponible en este entorno');
  }

  const randomBytes = new Uint8Array(length);
  cryptoObj.getRandomValues(randomBytes);
  const codeVerifier = toBase64Url(randomBytes).slice(0, length);

  const encoder = new TextEncoder();
  const data = encoder.encode(codeVerifier);
  const digest = await cryptoObj.subtle.digest('SHA-256', data);
  const codeChallenge = toBase64Url(new Uint8Array(digest));

  return {
    codeVerifier,
    codeChallenge,
  };
}

File: sdk/go/pkce.go
package peakauth

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
)

// GeneratePKCE crea un par code_verifier y code_challenge (S256) criptográficamente seguro.
// El tamaño por defecto es de 64 bytes (longitud típica recomendada por RFC 7636).
func GeneratePKCE(length ...int) (*PKCEPair, error) {
	size := 64
	if len(length) > 0 && length[0] >= 43 && length[0] <= 128 {
		size = length[0]
	}

	randomBytes := make([]byte, size)
	if _, err := rand.Read(randomBytes); err != nil {
		return nil, fmt.Errorf("error generando bytes aleatorios: %w", err)
	}

	verifier := base64.RawURLEncoding.EncodeToString(randomBytes)
	if len(verifier) > 128 {
		verifier = verifier[:128]
	}

	hash := sha256.Sum256([]byte(verifier))
	challenge := base64.RawURLEncoding.EncodeToString(hash[:])

	return &PKCEPair{
		CodeVerifier:  verifier,
		CodeChallenge: challenge,
	}, nil
}
